core/src/main/java/org/apache/iceberg/io/datafile/DataFileServiceRegistry.java

```diff
@@ -0,0 +1,420 @@
+        Map<Integer, ?> idToConstant,
+        DeleteFilter<?> deleteFilter);
+  }
+
+  /**
+   * Service building writers. Implementations should be registered through the {@link
+   * java.util.ServiceLoader}. {@link DataFileServiceRegistry} is used to collect and serve the
+   * writer implementations.
+   */
+  public interface WriterService<S> {
+    /** Returns the file format which is handled by the service. */
+    Key key();
+
+    /**
+     * Provides an appender builder for the given output file which writes objects with a given
+     * input type.
+     *
+     * @param outputFile to write to
+     * @param rowType of the input records
+     */
+    AppenderBuilder appenderBuilder(EncryptedOutputFile outputFile, S rowType);
+
+    /**
+     * Provides a data writer builder for the given output file which writes objects with a given
+     * input type.
+     *
+     * @param outputFile to write to
+     * @param rowType of the input records
+     */
+    DataWriterBuilder dataWriterBuilder(EncryptedOutputFile outputFile, S rowType);
+
+    /**
+     * Provides an equality delete writer builder for the given output file which writes objects
+     * with a given input type.
+     *
+     * @param outputFile to write to
+     * @param rowType of the input records
+     */
+    <T extends EqualityDeleteWriterBuilder<T>>
+        EqualityDeleteWriterBuilder<T> equalityDeleteWriterBuilder(
+            EncryptedOutputFile outputFile, S rowType);
+
+    /**
+     * Provides a positional delete writer builder for the given output file which writes objects
+     * with a given input type.
+     *
+     * @param outputFile to write to
+     * @param rowType of the input records
+     */
+    <T extends PositionDeleteWriterBuilder<T>>
+        PositionDeleteWriterBuilder<T> positionDeleteWriterBuilder(
+            EncryptedOutputFile outputFile, S rowType);
+  }
+
+  /** Key used to identify readers and writers in the {@link DataFileServiceRegistry}. */
+  public static class Key {
+    private final FileFormat fileFormat;
+    private final String dataType;
+    private final String builderType;
+
+    /**
+     * Create the key when there are no multiple builder types available.
+     *
+     * @param fileFormat the service handles
+     * @param dataType the service handles
+     */
+    public Key(FileFormat fileFormat, String dataType) {
+      this.fileFormat = fileFormat;
+      this.dataType = dataType;
+      this.builderType = null;
+    }
+
+    /**
+     * Create the key when there are multiple builder types available.
+     *
+     * @param fileFormat the service handles
+     * @param dataType the service handles
+     * @param builderType identifier of the builder type when multiple readers/writers are available
+     */
+    public Key(FileFormat fileFormat, String dataType, String builderType) {
+      this.fileFormat = fileFormat;
+      this.dataType = dataType;
+      this.builderType = builderType;
```

Comment:
I was thinking of defining the default one using a priority (`int`) based approach and let the one with the highest priority be the default one. WDYT?

Reply:
+1 for simplicity. This code should not determine things like whether Comet is used. This should have a single purpose, which is to standardize how object models plug in.

Reply:
Moved the config to properties, and the builder method will create the different readers based on this config